### ISSUE TITLE
Pkg3: "Binary" or "non-source" packages

### ISSUE THREAD

**Issue by tknopp** [JuliaLang/Juleps#23]
Taken from https://github.com/JuliaLang/julia/issues/16330:

It would be nice if it would be possible to have packages where the functionality is only available in precompiled form.

**Comment by tknopp** [comment]
Since this requires infrastructure within Julia Base/Core this does not necessary need to be implemented for the first Pkg3 release but it would be good if this feature could be kept in mind.

**Comment by samo-lin** [comment]
Are there any news on this issue? This feature would be most useful to avoid file system congestion when [running a julia application at large scale (on thousands of nodes) - a Julia discourse topic](https://discourse.julialang.org/t/run-a-julia-application-at-large-scale-on-thousands-of-nodes/23873/8).

**Comment by StefanKarpinski** [comment]
Not really. It requires non-trivial compiler work and hasn't been a top priority for the compiler team.

**Comment by nico202** [comment]
@tknopp 
> It would be nice

Nice to whom? It mean shipping non-free software. It's not nice not to have the source code. Please don't

**Comment by tknopp** [comment]
Yes, but still it would be nice if Pkg would not be restricted to open source packages. If one wants companies to adopt Julia, the possibility of distributing an obfuscated package should be given. You don't need to use that package but closed source software is to evil by definition.

**Comment by nico202** [comment]
If companies want to adopt julia, they need to release the source code. That's the deal. This is what we should promote

**Comment by tknopp** [comment]
That might be your personal opinion (I don't know who you refer to "we") but that is certainly not common sense across Julia developers. Julia itself is not licensed under GPL for a reason, its released under MIT, which allows for using the code in a closed source product. Most packages are also released under MIT license.

**Comment by nico202** [comment]
Yeah, I usually state my opinions, non somebody else's. Including Julia in a proprietary program is one thing. Allowing julia to download proprietary programs is another.

**Comment by andyferris** [comment]
> Allowing julia to download proprietary programs is another.

I mean... users can already just wrap a compiled library with `ccall`... there's no reason we wouldn't accept a wrapper library in `General` that wrapped closed-source (but distributable) code. I just found an old library for DirectX, but I'm sure there are better examples. Whether that compiled code was originally Julia code or not should be immaterial.
